### ISSUE TITLE
Applying NumpyPreproc to mne.Epochs

### ISSUE THREAD

**Issue by hubertjb** [braindecode/braindecode#160]
`NumpyPreproc` relies on the Raw/Epochs object to have an `apply_function` method, but it looks like `mne.Epochs` doesn't have one, meaning the preprocessor fails. I think we used to have a different pattern before that worked with Epochs too @gemeinl ?

Some ideas:
1- We could go back to the previous implementation of preprocessors, or at least part of it.
2- We see whether there could be a new `apply_function` method to `mne.Epochs` in MNE. @agramfort 
3- We handle the preprocessing differently for Raw and for Epochs objects.

**Comment by agramfort** [comment]
adding an apply_function method to Epochs is doable in mne

>

**Comment by robintibor** [comment]
Ok then to me it is fine if, until then our NumpyPreproc simply only works on raws... we could try  catch it if it is being applied to an epochs object to give a meaningful error message.
But of course if we had something that works on epochs before, could be interesting. Do you remember it @gemeinl ?

**Comment by gemeinl** [comment]
In the code, we have the following cases:
-1: MNEPreproc, Raw / Epochs, string     -> assumes Raw / Epochs has a function called string
-2: MNEPreproc, Raw / Epochs, callable  -> exposes Raw / Epochs to callable
-3: NumpyPreproc, Raw / Epochs, string  -> assumed Raw / Epochs has a function called string
-4: NumpyPreproc Raw, callable               -> uses 'apply_function' to apply callable to Raw
So basically, we would like mne to do the work for us (like modifying data in-place for Epochs), however, you could do it yourself with MNEPreproc and a callable. You should have full access to the object there.

**Comment by gemeinl** [comment]
Maybe this API is too complex / confusing. I already suggested to do some renaming here, we could discuss this again.
Also, please correct me if I am mistaken @hubertjb and if the second case does not work for you.

**Comment by hubertjb** [comment]
Thanks for the clarification @gemeinl! I hadn't realized that `MNEPreproc` and a callable does work in the case that I was working on.

In terms of API, I think it could be interesting to combine the two preprocessors into a single one. Is there a use case for which having the choice between `MNEPreproc` and NumpyPreproc` could be useful? Could your third case be merged with case 1?

Maybe adding a `apply_function` to `mne.Epochs` would make this even simpler on our side.

**Comment by robintibor** [comment]
Reason for NumpyPreproc is that code gets more concise in a lot of examples. I would like to keep it for that reason, that's why it was introduced. Else you have to supply apply function as str to MNEpreproc, then your numpy function, and if this has further arguments, have to wrap them in partial again etc. This made quite some code uglier before in my view, that's why NumpyPreproc is there.

**Comment by robintibor** [comment]
We could think about showcasing both ways of MNEPreproc (str and callable) in the main example to make it more discoverable

**Comment by hubertjb** [comment]
I see @robintibor. If mne.Epochs had an `apply_function` method, would there still be a use case for directly applying a callable to `raw_or_epochs.load_data()`? Because if not maybe a unified approach like could work:

```
if fn is callable:
    use apply_function
elif fn is a string:
    try to use method 'fn' of Raw/Epochs object
```

**Comment by hubertjb** [comment]
@robintibor just checking whether the approach just above could work? If so I can look into an MNE PR for `apply_function`.

**Comment by SulemanRasheed** [comment]
In case someone is trying to follow tutorials on the documentation page and having an issue while using the `NumpyPreproc` function with Epochs to implement functionality like _exponential_moving_standardization_ then write a custom function on `mne.epochs` and use 'MNEPreproc' to achieve the same functionality

Following worked for me
```
def custom_exp_moving_std_fn(epochs, factor_new=factor_new, init_block_size=init_block_size):
    data = epochs.get_data()
    for i in range(len(data)):
        epochs._data[i] = exponential_moving_standardize(data[i], 
                        factor_new=factor_new, init_block_size=init_block_size)
    return epochs

```
and use it like 
```
MNEPreproc(fn=custom_exp_moving_std_fn, factor_new=factor_new,
        init_block_size=init_block_size)
```

**Comment by robintibor** [comment]
Outdated
